shard/committee/assignment.go

```diff
@@ -35,11 +34,134 @@ type Reader interface {
+}
+
+// CompletedEPoSRound ..
+type CompletedEPoSRound struct {
+	MedianStake         numeric.Dec              `json:"epos-median-stake"`
+	MaximumExternalSlot int                      `json:"max-external-slots"`
+	AuctionWinners      []effective.SlotPurchase `json:"epos-slot-winners"`
+	AuctionCandidates   []*CandidateOrder        `json:"epos-slot-candidates"`
+}
+
+// CandidateOrder ..
+type CandidateOrder struct {
+	*effective.SlotOrder
+	Validator common.Address
+}
+
+// MarshalJSON ..
+func (p CandidateOrder) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		*effective.SlotOrder
+		Validator string `json:"validator"`
+	}{
+		p.SlotOrder, common2.MustAddressToBech32(p.Validator),
+	})
+}
+
+// NewEPoSRound runs a fresh computation of EPoS using
+// latest data always
+func NewEPoSRound(stakedReader StakingCandidatesReader) (
+	*CompletedEPoSRound, error,
+) {
+	eligibleCandidate, err := prepareOrders(stakedReader)
+	if err != nil {
+		return nil, err
+	}
+	maxExternalSlots := shard.ExternalSlotsAvailableForEpoch(
+		stakedReader.CurrentBlock().Epoch(),
+	)
+	median, winners := effective.Apply(
+		eligibleCandidate, maxExternalSlots,
+	)
+	auctionCandidates := make([]*CandidateOrder, len(eligibleCandidate))
+
+	i := 0
+	for key := range eligibleCandidate {
+		auctionCandidates[i] = &CandidateOrder{
+			SlotOrder: eligibleCandidate[key],
+			Validator: key,
+		}
+		i++
+	}
+
+	return &CompletedEPoSRound{
+		MedianStake: median,
+		// NOTE + 1 because we use it as index in assigment,
+		// but as count in semantics
+		MaximumExternalSlot: maxExternalSlots + 1,
+		AuctionWinners:      winners,
+		AuctionCandidates:   auctionCandidates,
+	}, nil
+}
+
+func prepareOrders(
+	stakedReader StakingCandidatesReader,
+) (map[common.Address]*effective.SlotOrder, error) {
+	candidates := stakedReader.ValidatorCandidates()
+	blsKeys := map[shard.BlsPublicKey]struct{}{}
+	essentials := map[common.Address]*effective.SlotOrder{}
+	totalStaked, tempZero := big.NewInt(0), numeric.ZeroDec()
+
+	for i := range candidates {
+		validator, err := stakedReader.ReadValidatorInformation(
+			candidates[i],
+		)
+		if err != nil {
+			return nil, err
+		}
+		if !staking.IsEligibleForEPoSAuction(validator) {
+			continue
+		}
+
+		validatorStake := big.NewInt(0)
+		for i := range validator.Delegations {
+			validatorStake.Add(
+				validatorStake, validator.Delegations[i].Amount,
+			)
+		}
+
+		totalStaked.Add(totalStaked, validatorStake)
+
+		found := false
+		for _, key := range validator.SlotPubKeys {
+			if _, ok := blsKeys[key]; ok {
+				found = true
+			} else {
+				blsKeys[key] = struct{}{}
+			}
+		}
+
+		if found {
+			continue
+		}
+
+		essentials[validator.Address] = &effective.SlotOrder{
+			validatorStake,
+			validator.SlotPubKeys,
+			tempZero,
+		}
+	}
+	totalStakedDec := numeric.NewDecFromBigInt(totalStaked)
+
+	for _, value := range essentials {
+		value.Percentage = numeric.NewDecFromBigInt(value.Stake).Quo(totalStakedDec)
```

Comment:
What is this Percentage used for ?

Reply:
because the function prepares orders, people downstream been asking in epos, what % of the total state did I have during auction, so that median-stake RPC gets more feature rich: 

https://gist.github.com/fxfactorial/6c518ac0e90c3978f8420d414c3cf611#file-median-stake-json